refactor(common): drop unnecessary `srcset` sanitization

AndrewKushnir opened the pull request:
This commit updates runtime and compiler to drop unnecessary `srcset` sanitization. The sanitization was needed previously for old browsers, but all modern browsers can handle `srcset` safely without any additional sanitization.

See prior discussion in https://github.com/angular/angular/pull/45182.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

bjarkler commented on packages/compiler/src/schema/dom_element_schema_registry.ts:
```diff
@@ -117,7 +116,7 @@ const SCHEMA: string[] = [
   'h1,h2,h3,h4,h5,h6^[HTMLElement]|align',
   'html^[HTMLElement]|version',
   'iframe^[HTMLElement]|align,!allowFullscreen,frameBorder,height,longDesc,marginHeight,marginWidth,name,referrerPolicy,%sandbox,scrolling,src,srcdoc,width',
-  'img^[HTMLElement]|align,alt,border,%crossOrigin,#height,#hspace,!isMap,longDesc,lowsrc,name,referrerPolicy,sizes,src,srcset,useMap,#vspace,#width',
```
Correct me if I'm wrong but I don't believe you should be removing srcset from the DOM schema, since it's still a valid attribute for img and source.

AndrewKushnir replied:
Yes, great catch! I've reverted this change.

bjarkler commented on packages/core/src/sanitization/html_sanitizer.ts:
```diff
@@ -64,9 +65,6 @@ export const VALID_ELEMENTS =
 export const URI_ATTRS = tagSet('background,cite,href,itemtype,longdesc,poster,src,xlink:href');
 
-// Attributes that have special href set hence need to be sanitized
-export const SRCSET_ATTRS = tagSet('srcset');
```
I believe srcset should be moved to HTML_ATTRS instead. If you remove it entirely from VALID_ATTRS below, the sanitizer will always drop srcset as an unrecognized attribute.

AndrewKushnir replied:
👍

bjarkler commented on packages/core/test/sanitization/html_sanitizer_spec.ts:
```diff
@@ -69,11 +69,6 @@ function sanitizeHtml(defaultDoc: any, unsafeHtmlInput: string): string {
 
-    it('sanitizes srcset attributes', () => {
-      expect(sanitizeHtml(defaultDoc, '<img srcset="/foo.png 400px, javascript:evil() 23px">'))
-          .toEqual('<img srcset="/foo.png 400px, unsafe:javascript:evil() 23px">');
```
Consider keeping this test and updating it to ensure that srcset is in fact not sanitized. I believe currently the srcset attribute would be dropped completely by the sanitizer, even if the URL is benign.

AndrewKushnir replied:
I've updated this test to illustrate that there are no changes to the `srcset` value are performed.

AndrewKushnir committed:
refactor(common): drop unnecessary srcset sanitization

This commit updates runtime and compiler to drop unnecessary `srcset` sanitization. The sanitization was needed previously for old browsers, but all modern browsers can handle `srcset` safely without any additional sanitization.

See prior discussion in https://github.com/angular/angular/pull/45182.

Resolves #45164.

AndrewKushnir commented on packages/compiler/src/schema/dom_security_schema.ts:
```diff
@@ -35,10 +35,24 @@ export function SECURITY_SCHEMA(): {[k: string]: SecurityContext} {
-      'a|ping',       'blockquote|cite', 'body|background', 'del|cite',     'form|action',
-      'img|src',      'img|srcset',      'input|src',       'ins|cite',     'q|cite',
-      'source|src',   'source|srcset',   'track|src',       'video|poster', 'video|src',
+      '*|formAction',
```
Note to reviewers: this list was auto-formatted, the only change is dropping the `img|srcset` and `source|srcset` entries.

bjarkler commented on packages/core/test/sanitization/html_sanitizer_spec.ts:
```diff
@@ -69,9 +69,10 @@ function sanitizeHtml(defaultDoc: any, unsafeHtmlInput: string): string {
     });
 
-    it('sanitizes srcset attributes', () => {
+    it('ignores srcset attributes', () => {
```
nit: "allows srcset attributes" may be a bit more clear

bjarkler reviewed (approved):
Looks great!

pkozlowski-opensource reviewed (commented):
The change LGTM, thnx @AndrewKushnir for this cleanup! 

I was wondering if it would make sense to add a test covering the use-case from #45164 and mark this PR as fixing #45164? Or do you want to take care of #45164 (add a dedicated test) in a separate PR?

AndrewKushnir commented:
Thanks for the feedback @pkozlowski-opensource 👍 I've added a test based on https://github.com/angular/angular/issues/45164 and updated the commit message to auto-close that ticket on merge.

pkozlowski-opensource reviewed (approved):
Thnx for adding the test covering https://github.com/angular/angular/pull/45182 @AndrewKushnir , this LGTM!

Reviewed-for: fw-core 
Reviewed-for: fw-security

jelbourn reviewed (approved):
LGTM

Reviewed-for: fw-security

dylhunn commented:
This PR was merged into the repository by commit 9e2d3ed2d13c0e12bc2e7ae733661ac2c2e97fed.
